### PR TITLE
[ENH] Categorization of Age and Sex Columns #20

### LINKED ISSUE #20: categorization of `age`and `sex`
- [x] Assigning category labels "Age"  and "Sex " to columns based on the column header and entries.
- [ ] Taking care of missing values.
- [x] Ensuring that the LLM is aware of most possible inputs.(As there are slight variations in how these data is recorded during different studies)

### PR DESCRIPTION
**Pull request by tvilight4**
- Closes #20 

Changes proposed in this pull request:

- Added the functions to return the levels for Sex and Format for Age
- The output returned is in accordance to the requirements of the parsing code.
-  The test file has been added to visualize what the codebase returns.

## Checklist
_This section is for the PR reviewer_

- [x] PR has an interpretable title with a prefix (`[ENH]`, `[FIX]`, `[REF]`, `[TST]`, `[CI]`, `[MNT]`, `[INF]`, `[MODEL]`, `[DOC]`) _(see our [Contributing Guidelines](https://neurobagel.org/contributing/CONTRIBUTING#pull-request-guidelines) for more info)_
- [ ] PR has a label for the release changelog or `skip-release` (to be applied by maintainers only)
- [ ] PR links to GitHub issue with mention `Closes #XXXX`
- [ ] Tests pass
- [ ] Checks pass

For new features:
- [ ] Tests have been added

For bug fixes:
- [ ] There is at least one test that would fail under the original bug conditions.

### CODE REVIEW

File: categorization/bin/llm_categorization.py [hunk 1]
```diff
@@ -1,22 +1,121 @@
...
+        output: Dict[str, Union[str, Dict[str, str]]] = (
+            {}
+        )  # Or any default output as per your requirement
+        print(json.dumps(output))
+
+    output = {
+        "TermURL": "nb:Sex",
+        "Levels": {str(var1): "male", str(var2): "female"},
+    }
+    print(json.dumps(output))
```
Comment (barbarastrasser) [inline categorization/bin/llm_categorization.py:38]:
```suggestion
    return output
```

File: categorization/bin/llm_categorization.py [hunk 2]
```diff
@@ -1,22 +1,121 @@
...
+        elif is_european_decimal(num_str):
+            Fvar = "europeanDecimalValue"
+            break
+        # 2 more conditions yet to be added
+
+    output: Dict[str, str] = {
+        "TermURL": "nb:Sex",
+        "Format": Fvar,
+    }
+    print(json.dumps(output))
```
Comment (barbarastrasser) [inline categorization/bin/llm_categorization.py:104]:
```suggestion
    return output
```

File: categorization/bin/llm_categorization.py [hunk 3]
```diff
@@ -56,12 +155,10 @@ def llm_invocation(result_dict: Dict[str, str]) -> Dict[str, Any]:
     elif "Session_IDs" in r:
         output = {"TermURL": "nb:Session"}
         print(f"{json.dumps(output)}")
-    elif "Age" in r:
-        output = {"TermURL": "nb:Age"}
-        print(f"{json.dumps(output)}")
     elif "Sex" in r:
-        output = {"TermURL": "nb:Sex"}
-        print(f"{json.dumps(output)}")
+        SexLevel(result_dict, r, key)
```
Comment (barbarastrasser) [inline categorization/bin/llm_categorization.py:159]:
```suggestion
        output = SexLevel(result_dict, r, key)
```

File: categorization/bin/llm_categorization.py [hunk 4]
```diff
@@ -56,12 +155,10 @@ def llm_invocation(result_dict: Dict[str, str]) -> Dict[str, Any]:
...
+    elif "Age" in r:
+        AgeFormat(result_dict, r, key)
```
Comment (barbarastrasser) [inline categorization/bin/llm_categorization.py:161]:
```suggestion
        output = AgeFormat(result_dict, r, key)
```

File: categorization/bin/llm_categorization.py [hunk 5]
```diff
@@ -1,22 +1,121 @@
...
+        elif is_iso8601(num_str):
+            Fvar = "ISO"
+            break
+        elif is_european_decimal(num_str):
+            Fvar = "europeanDecimalValue"
+            break
+        # 2 more conditions yet to be added
+
+    output: Dict[str, str] = {
+        "TermURL": "nb:Sex",
```
Comment (barbarastrasser) [inline categorization/bin/llm_categorization.py:101]:
```suggestion
        "TermURL": "nb:Age",
```
That's important ;)

File: categorization/bin/llm_categorization.py [hunk 6]
```diff
@@ -1,22 +1,121 @@
...
+            break
+        elif is_float(num_str):
+            Fvar = "FloatVal"
+            break
+        elif is_iso8601(num_str):
+            Fvar = "ISO"
+            break
+        elif is_european_decimal(num_str):
+            Fvar = "europeanDecimalValue"
+            break
```
Comment (barbarastrasser) [inline categorization/bin/llm_categorization.py:97]:
```suggestion
        if is_integer(num_str):
            Fvar = "integervalue"
            break
        elif is_float(num_str):
            Fvar = "floatvalue"
            break
        elif is_iso8601(num_str):
            Fvar = "iso8601"
            break
        elif is_european_decimal(num_str):
            Fvar = "europeandecimalvalue"
            break
```
just to match the strings needed for correct parsing

**Review by barbarastrasser — CHANGES_REQUESTED** [review]
I suggested some changes to the code since I ran into `Error message: local variable 'output' referenced before assignment` ;) Additionally, I made some general changes (terms, etc..) With the changes it works fine for me - maybe you can try it out.

File: categorization/bin/llm_categorization.py [hunk 7]
```diff
@@ -1,22 +1,121 @@
...
+        var2 = "1"
+    elif "M" in value or "F" in value:
+        var1 = "M"
+        var2 = "F"
+    elif "m" in value or "f" in value:
+        var1 = "m"
+        var2 = "f"
+    elif "male" in value or "female" in value:
+        var1 = "male"
+        var2 = "female"
```
Comment (barbarastrasser) [inline categorization/bin/llm_categorization.py:27]: Did you include something like `.lower()` in the code for the sex levels? If so, nevermind - if not - maybe we should to capture all the valid formats of the [BIDS sex column specification](https://bids-specification.readthedocs.io/en/stable/glossary.html#objects.columns.sex). And maybe include the "Other" parts also already? - I should include this in the parsing also ;)

Comment (barbarastrasser) [inline categorization/bin/llm_categorization.py:27]:
```suggestion
    var1: str = ""
    var2: str = ""
    var3: str = ""

    if "1" in value or "2" in value or "3" in value:
        var1 = "1"
        var2 = "2"
        var3 = "3"
    elif "0" in value or "1" in value or "2" in value:
        var1 = "0"
        var2 = "1"
        var3 = "2"
    elif "M" in value or "F" in value or "O" in value:
        var1 = "M"
        var2 = "F"
        var3 = "O"
    elif "m" in value or "f" in value or "o" in value:
        var1 = "m"
        var2 = "f"
        var3 = "o"
    elif "male" in value or "female" in value or "other" in value:
        var1 = "male"
        var2 = "female"
        var3 = "other"
```

Comment (tvilight4) [reply categorization/bin/llm_categorization.py]: Yes .lower() will be better!! :smile: 
I will add it and the others section and ping you!

**Review by barbarastrasser — APPROVED** [review]
The code runs perfectly fine for me now :) Feel free to merge in :)

**Pull request merged**